src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalPackageSearchResource.cs

```diff
@@ -59,6 +59,21 @@ public async override Task<IEnumerable<IPackageSearchMetadata>> SearchAsync(
                     query = query.Where(package => ContainsAnyTerm(terms, package));
                 }
 
+                // Filter on package types
+                if (filters?.PackageTypes != null
+                    && filters.PackageTypes.Any())
+                {
+                    foreach (var packageTypeName in filters.PackageTypes)
+                    {
+                        query = query
+                            .Where(package => package.Nuspec
+                                .GetPackageTypes()
+                                .Any(packageType => StringComparer.OrdinalIgnoreCase.Equals(
+                                    packageType.Name,
+                                    packageTypeName)));
```

Comment:
As per the PR checklist, please add tests. There's already a test class named [`LocalpackageSearchResourceTests`](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs)

Reply:
@zivkan I would love to, but I am blocked setting up a working full environment due to package restore issues:

```
Failed to download package 'Microsoft.VisualStudio.Sdk.17.10.40171' from 'https://pkgs.dev.azure.com/azure-public/3ccf6661-f8ce-4e8a-bb2e-eff943ddd3c7/_packaging/36a629e1-6c5b-4bcd-aa2e-6018802d6b99@2f2420a2-f975-407f-ad27-d1e0f85b9b15/nuget/v3/flat2/microsoft.visualstudio.sdk/17.10.40171/microsoft.visualstudio.sdk.17.10.40171.nupkg'.
  An error occurred while sending the request.
    The underlying connection was closed: An unexpected error occurred on a send.
    Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host.
    An existing connection was forcibly closed by the remote host
```

I can restore nuget packages successfully on my machine and have confirmed all TLS mitigation strategies are in place as per https://github.com/microsoft/azure-devops-tls12 and mentioned in https://github.com/NuGet/NuGetGallery/issues/9121.

I will continue investigating, but if I am missing anything obvious, please advise.

Reply:
This is extremely weird, but solved this problem by forcing IPv4 using https://community.fabric.microsoft.com/t5/Power-Query/The-underlying-connection-was-closed-An-unexpected-error/m-p/3442332#M112277

Working on tests now, but just leaving this here as it might indicate some kind of possible TLS issue with IPv6 on some machines / connections that was hard to find / diagnose.

Reply:
@zivkan One last API design question which just occurred to me: what should the behavior of `LocalPackageSearchResource` be if the package type `Dependency` is specified and the local package does not specify package types, e.g. for older packages without the `packageTypes` metadata?

In [the docs](https://learn.microsoft.com/en-us/nuget/create-packages/set-package-type?tabs=dotnet#known-package-types) it is stated that _"Packages not marked with a type, including all packages created with earlier versions of NuGet, default to the `Dependency` type."_

This would suggest that we should accept matches for all packages without `packageTypes` metadata if the specified package type filter is `Dependency`.

Reply:
sorry for the delay in replying. Too much work and not enough time.

> This would suggest that we should accept matches for all packages without packageTypes metadata if the specified package type filter is Dependency.

This is my preference. I'd rather `GetPackageTypes` return what the package's nuspec actually reports, just in case there are scenarios where the difference between `Dependency` not "not specified" is important elsewhere.

Reply:
Fixed in https://github.com/NuGet/NuGet.Client/pull/5991/commits/9b1d7696c91638e653566f66dd70a4f27a6f65af